File: elixir-phoenix/lib/rocket/store/bootstrap.ex
defmodule Rocket.Store.Bootstrap do
  @moduledoc "Creates system tables and seeds the admin user."

  alias Rocket.Store

  def bootstrap(conn) do
    dialect = Store.dialect()

    dialect.system_tables_sql()
    |> String.split(";")
    |> Enum.map(&String.trim/1)
    |> Enum.reject(&(&1 == ""))
    |> Enum.each(fn sql ->
      Store.exec!(conn, sql)
    end)

    seed_admin_user(conn, dialect)
  end

  defp seed_admin_user(conn, dialect) do
    case Store.query_row(conn, "SELECT COUNT(*) as count FROM _users") do
      {:ok, %{"count" => count}} when count == 0 or count == false ->
        hash = Bcrypt.hash_pwd_salt("changeme")
        roles = dialect.array_param(["admin"])

        {sql, params} =
          if dialect.uuid_default() == "" do
            # SQLite: no DEFAULT gen_random_uuid(), generate UUID manually
            id = Ecto.UUID.generate()

            {"INSERT INTO _users (id, email, password_hash, roles) VALUES ($1, $2, $3, $4)",
             [id, "admin@localhost", hash, roles]}
          else
            {"INSERT INTO _users (email, password_hash, roles) VALUES ($1, $2, $3)",
             ["admin@localhost", hash, roles]}
          end

        Store.exec!(conn, sql, params)

        require Logger

        Logger.warning(
          "Default admin user created (admin@localhost / changeme) — change the password immediately."
        )

      _ ->
        :ok
    end
  end
end

File: backend/expressjs/src/store/bootstrap.ts
import crypto from "node:crypto";
import bcrypt from "bcryptjs";
import { getDialect, SQLiteDatabase } from "./postgres.js";
import type { Queryable } from "./postgres.js";

export async function bootstrap(pool: Queryable): Promise<void> {
  const ddl = getDialect().systemTablesSQL();

  if (pool instanceof SQLiteDatabase) {
    pool.execMulti(ddl);
  } else {
    await (pool as any).query(ddl);
  }

  await seedAdminUser(pool);
}

async function seedAdminUser(pool: Queryable): Promise<void> {
  const result = pool instanceof SQLiteDatabase
    ? pool.query("SELECT COUNT(*) AS count FROM _users")
    : await (pool as any).query("SELECT COUNT(*) FROM _users");
  const count = parseInt(result.rows[0].count, 10);
  if (count > 0) return;

  const hash = await bcrypt.hash("changeme", 10);
  const dialect = getDialect();

  if (dialect.name() === "sqlite") {
    const id = crypto.randomUUID();
    (pool as SQLiteDatabase).query(
      "INSERT INTO _users (id, email, password_hash, roles) VALUES ($1, $2, $3, $4)",
      [id, "admin@localhost", hash, dialect.arrayParam(["admin"])],
    );
  } else {
    await (pool as any).query(
      "INSERT INTO _users (email, password_hash, roles) VALUES ($1, $2, $3)",
      ["admin@localhost", hash, ["admin"]],
    );
  }

  console.log(
    "WARNING: Default admin user created (admin@localhost / changeme) — change the password immediately.",
  );
}
